[Dashboard] Fix potential job actor leak by ensuring retrieval of actor instance before termination.

daiping8 opened the pull request:
## Why are these changes needed?

To close issue #55858

## Problem Analysis

Let's analyze this from the code details. Let's look at the [code](https://github.com/ray-project/ray/blob/6fcc861fb04229b9dfe216531e47af3ea477afa2/python/ray/dashboard/modules/job/job_manager.py#L168-L346) in `job_manager.py` that handles job pending timeouts by the loop. 

### Normal process

When a job times out, JobStatus will be FAILED and the entire loop will be break. At the end of the function, `ray.kill(job_supervisor, no_restart=True)` will be executed.


### Faulty process

When a job is in the pending state, for some reason the old node that submitted the job dies, and a new node is created . For the new job manager created after the  raylet on new node **(new raylet)** starts, it continuously monitors the task status through `_monitor_job_internal`. 

Just like the scenario I mentioned in #55858 (a new raylet is in the process of starting up, and the job happens to reach its timeout during this period), the loop that checks for job pending timeouts terminates **during its first iteration**, so the [code](https://github.com/ray-project/ray/blob/6fcc861fb04229b9dfe216531e47af3ea477afa2/python/ray/dashboard/modules/job/job_manager.py#L233-L234) responsible for obtaining the handle is never executed. It fails to obtain the `job_supervisor` handle.

So `ray.kill(job_supervisor, no_restart=True)` will not be executed.


## Solution

- Fix potential job actor leak by ensuring retrieval of actor instance before termination.
- Add a test to validate job timeout behavior during new raylet creation.

## Related issue number

Closes #55858

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Ensures `_monitor_job_internal` fetches the job supervisor actor before killing it to avoid leaks, and adds a test covering pending-job timeout during new raylet creation.
> 
> - **Jobs Manager (`python/ray/dashboard/modules/job/job_manager.py`)**:
>   - Ensure cleanup: before killing, fetch `job_supervisor` via `_get_actor_for_job(job_id)` if missing to avoid actor leaks on early loop exit.
> - **Tests (`python/ray/dashboard/modules/job/tests/test_job_manager.py`)**:
>   - Add `test_pending_job_timeout_during_raylet_creation` to simulate new raylet creation, verify pending job times out, and confirm supervisor actor is removed.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5274288ba69be100fbc5d9099507de246907371c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

daiping8 committed:
- Fix potential actor leak by initializing `job_supervisor` in defensive kill logic.
- Add test to verify timeout handling for pending jobs during new raylet creation.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 commented:
@jjyao Thank you for your attention. I have clarified the relevant content in the descriptions of the issue and PR. The core problem is that the `job_supervisor` handle can be `None` in specific scenarios, thereby preventing resource release.

This scenarios is an extreme edge case, which makes it difficult to describe and understand. Thank you again for your patience.

jjyao commented:
@daiping8 could you rebase, I just merged my fix.

daiping8 committed:
Add test for pending job timeout during raylet creation

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Merge branch 'master' into job_sub

daiping8 committed:
Refactor: Adjust environment variable setting for job timeout test

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Merge branch 'job_sub' of https://github.com/daiping8/ray into job_sub

daiping8 committed:
Merge branch 'master' into job_sub

jjyao commented on python/ray/dashboard/modules/job/tests/test_job_manager.py:
```diff
@@ -353,6 +353,58 @@ async def test_runtime_env_setup_logged_to_job_driver_logs(
+async def test_pending_job_timeout_during_raylet_creation(
+    call_ray_start, tmp_path, monkeypatch  # noqa: F811
+):
+    """Test the timeout for pending jobs during new raylet creation."""
```
Do you mean head node here?

jjyao commented on python/ray/dashboard/modules/job/tests/test_job_manager.py:
```diff
@@ -353,6 +353,58 @@ async def test_runtime_env_setup_logged_to_job_driver_logs(
+
+    # Submit a job with unsatisfied resource.
+    job_id = "test_job_1"
+    # Skip '_monitor_job' in 'submit_job', to avoid affecting the simulation of new raylet.
+    job_manager.monitored_jobs.add(job_id)
```
This is too dependent on the internal implementation of the code: i.e. it depends on this specific code

```
async def _monitor_job(
        self, job_id: str, job_supervisor: Optional[ActorHandle] = None
    ):
        """Monitors the specified job until it enters a terminal state.

        This is necessary because we need to handle the case where the
        JobSupervisor dies unexpectedly.
        """
        if job_id in self.monitored_jobs:
            logger.debug(f"Job {job_id} is already being monitored.")
            return
```

jjyao commented on python/ray/dashboard/modules/job/tests/test_job_manager.py:
```diff
@@ -353,6 +353,58 @@ async def test_runtime_env_setup_logged_to_job_driver_logs(
+    ["ray start --head --num-cpus=1"],
+    indirect=True,
+)
+async def test_pending_job_timeout_during_raylet_creation(
```
I think what we want is to make the FIRST timeout check fails with timeout. I think the best way is to fake `time.time()` to return a time we want. 

Concretely we can pass a timer to JobManager that defaults to time.time() but in test, we can create a fake timer that returns a time we want. See https://github.com/ray-project/ray/blob/db1127bde72aeb9854adb7f408345b270a25e979/python/ray/serve/tests/unit/test_proxy_state.py#L199 as an example. This way we can make sure the first timeout check fails. Although this also depends on the internal implementation that time.time is called twice: first to get the start time of a job and the second time is to check if the job times out. I'm not sure if there is a better way cc @edoakes

edoakes replied:
This is the approach I would take

edoakes commented:
@daiping8 @jjyao there is really no reason for us to be running the `JobManager` on every node. This is an artifact of a half-finished feature to load balance job submission across nodes, which is not really necessary because we can still schedule drivers across the cluster while only having the job manager on the head node.

So my preferred fix for this would actually be to simplify things by running the `JobManager` only on the head node and removing the `_recover_running_jobs` logic entirely. What do you think?

daiping8 committed:
Update python/ray/dashboard/modules/job/tests/test_job_manager.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ping Dai <dai.ping88@zte.com.cn>

jjyao commented:
> So my preferred fix for this would actually be to simplify things by running the JobManager only on the head node and removing the _recover_running_jobs logic entirely. What do you think?

@edoakes yea, strictly speaking we don't officially support GCS FT for jobs, so we don't need `_recover_running_jobs` but in reality, many OSS users still run GCS FT for ray jobs and hence `_recover_running_jobs` was introduced in the first place.

daiping8 committed:
Remove test for pending job timeout during raylet creation as part of cleanup

Change-Id: I2786781cc39bc6af76e94318ce47160341f8b656
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Add `MockTimer` utility and test for pending job timeout during new head creation

- Introduced `MockTimer` class for simulating time manipulations in tests.
- Added test to validate timeout behavior for pending jobs during new head node creation.

Change-Id: Id13ea5bb41139bf0764ac3643b2e19a403a03a0b
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Refactor import order and add noqa directive for test function

Change-Id: Ie3962b7f8644367e7f9bf940baef828dbb02fb56
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
refactor(job): improve JobManager timing logic

- Introduced `Timer` class to replace direct `time.time()` calls, enhancing test maintainability.
- Refactored `__init__` to include a `timer` parameter for dependency injection.
- Updated `test_job_manager.py` tests to use `FakeTimer` instead of `MockTimer`.
- Renamed `MockTimer` to `FakeTimer` for better clarity of purpose.

Change-Id: Ia5351a1745cda42b3057e0ff70ccb02bd50cf538
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

jjyao commented on python/ray/dashboard/modules/job/job_manager.py:
```diff
@@ -35,6 +35,7 @@
 from ray.exceptions import ActorDiedError, ActorUnschedulableError, RuntimeEnvSetupError
 from ray.job_submission import JobErrorType, JobStatus
 from ray.runtime_env import RuntimeEnvConfig
+from ray.serve._private.utils import Timer
```
We should move Timer to `ray/_common/utils.py` so it can be shared by everyone

daiping8 committed:
refactor(utils): unify `Timer` and `TimerBase` into `_common.utils`

- Moved `Timer` and `TimerBase` from `serve._private.utils` to `_common.utils` for better reusability.
- Updated relevant imports across modules.
- Refactored `JobManager` to use `timeout_check_timer` for enhanced clarity.
- Adjusted related tests to align with new `Timer` implementation.

Change-Id: I28717c4ed0fe9322c29d03f799a7a200c8ee0aa3
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Merge branch 'ray-project:master' into job_sub

daiping8 committed:
Update `test_job_manager.py` to use `FakeTimer` for timeout handling

- Replaced `start_timer.advance` with `FakeTimer` instance in the test.
- Aligned job manager initialization to use the updated timer instance for consistency.

Change-Id: I4a7adac95e9e8575d51b171946fec8bf5cb8b466
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Merge branch 'job_sub' of https://github.com/daiping8/ray into job_sub

Change-Id: I7f304b4d1ed2e8f1bffd83be38a9a457569fbad6

jjyao commented on python/ray/dashboard/modules/job/job_manager.py:
```diff
@@ -70,14 +70,17 @@ class JobManager:
         self._cluster_id_hex = gcs_client.cluster_id.hex()
         self._log_client = JobLogStorageClient()
         self._supervisor_actor_cls = ray.remote(JobSupervisor)
+        self._timer = timeout_check_timer or Timer()
```
```suggestion
        self._timeout_check_timer = timeout_check_timer or Timer()
```

jjyao reviewed (approved):
Very sorry for the late review

jjyao committed:
Apply suggestions from code review

Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

jjyao committed:
Merge branch 'master' into job_sub

Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

jjyao commented on python/ray/dashboard/modules/job/job_manager.py:
```diff
@@ -185,7 +188,7 @@ async def _monitor_job_internal(
                         )
 
-                    if time.time() - job_info.start_time / 1000 > timeout:
+                    if self._timeout_check_timer.time() - job_info.start_time / 1000 > timeout:
```
```suggestion
                    if (
                        self._timeout_check_timer.time() - job_info.start_time / 1000 
                        > timeout
                    ):
```

jjyao commented on python/ray/dashboard/modules/job/tests/test_job_manager.py:
```diff
@@ -354,6 +356,60 @@ async def test_runtime_env_setup_logged_to_job_driver_logs(
+    timeout = DEFAULT_JOB_START_TIMEOUT_SECONDS
+    timeout_timer = FakeTimer(start_timer.time() + timeout + 1)
+
+    new_job_manager = JobManager(gcs_client, tmp_path, timeout_check_timer=timeout_timer)
```
```suggestion
    new_job_manager = JobManager(
        gcs_client, tmp_path, timeout_check_timer=timeout_timer
    )
```

jjyao committed:
Apply suggestions from code review

Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

daiping8 committed:
Merge branch 'master' of https://github.com/daiping8/ray into job_sub

Change-Id: Ice13a8aaf1db5a07b05c8f54a2eccbb55b91b3a2

daiping8 committed:
format

Change-Id: I1cdaae76d9db40441a7b7816f74a8764059dc54a
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

jjyao committed:
Merge branch 'master' into job_sub

daiping8 committed:
docs(ray-references): Adjusted the format of the id column to remove extra spaces

Change-Id: Ie1dddf8a96fd6b7cd8e4059f1597fd0f35765cdb
Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 commented:
@jjyao  All CI test have passed! Thanks.
